src/qibo/backends/__init__.py

```diff
@@ -0,0 +1,113 @@
+from qibo import config
+from qibo.config import raise_error, log, warnings, BACKEND_NAME
+from qibo.backends.numpy import NumpyBackend
+from qibo.backends.tensorflow import TensorflowBackend
+
+
+CONSTRUCTED_BACKENDS = {}
+def construct_backend(name):
+    if name not in CONSTRUCTED_BACKENDS:
+        if name == "numpy":
+            CONSTRUCTED_BACKENDS["numpy"] = NumpyBackend()
+        elif name == "tensorflow":
+            CONSTRUCTED_BACKENDS["tensorflow"] = TensorflowBackend()
+        else:
+            raise_error(ValueError, "Unknown backend name {}.".format(name))
+    return CONSTRUCTED_BACKENDS.get(name)
+
+numpy_backend = construct_backend("numpy")
+numpy_matrices = numpy_backend.matrices
+
+if BACKEND_NAME == "tensorflow":
+    K = TensorflowBackend()
+else: # pragma: no cover
+    # CI uses tensorflow as default backend
+    K = NumpyBackend()
+
+def set_backend(backend="custom"):
+    """Sets backend used to implement gates.
+
+    Args:
+        backend (str): possible options are 'custom' for the gates that use
+            custom tensorflow operator and 'defaulteinsum' or 'matmuleinsum'
+            for the gates that use tensorflow primitives (``tf.einsum`` or
+            ``tf.matmul`` respectively).
+    """
+    if not config.ALLOW_SWITCHERS and backend != K.gates:
+        warnings.warn("Backend should not be changed after allocating gates.",
+                      category=RuntimeWarning)
+
+    gate_backend = backend.split("_")
+    if len(gate_backend) == 1:
+        calc_backend, gate_backend = BACKEND_NAME, gate_backend[0]
+    elif len(gate_backend) == 2:
+        calc_backend, gate_backend = gate_backend
+    else:
+        raise_error(ValueError, "Unknown backend {}.".format(backend))
+    bk = construct_backend(calc_backend)
```

Comment:
As you did in `set_precision`, maybe we should keep both backends in sync, for example configurations:
- tensorflow: custom, numpy: defaulteinsum
- tensorflow & numpy: defaulteinsum
- tensorflow & numpy: matmuleinsum

What do you think?